restapi/src/main/java/io/stargate/web/resources/Converters.java

```diff
@@ -78,336 +82,762 @@
+              "Invalid UDT value: expected a JSON object or a string");
         }
+      default:
+        throw new AssertionError("Unsupported data type: " + type.rawType());
+    }
+  }
 
-        return udtValue;
+  private static <I> I toCqlInt(
+      Column.ColumnType type,
+      Object jsonValue,
+      Function<BigInteger, I> fromBigIntegerExact,
+      Function<Number, I> fromNumber,
+      long min,
+      long max) {
+    if (jsonValue instanceof BigInteger) {
+      try {
+        return fromBigIntegerExact.apply((BigInteger) jsonValue);
+      } catch (ArithmeticException e) {
+        throw new IllegalArgumentException(
+            String.format("Invalid %s value %s: out of range", type.rawType(), jsonValue));
+      }
     }
+    if (jsonValue instanceof Integer || jsonValue instanceof Long) {
+      Number number = (Number) jsonValue;
+      long longValue = number.longValue();
+      if (longValue < min || longValue > max) {
+        throw new IllegalArgumentException(
+            String.format("Invalid %s value %s: out of range", type.rawType(), jsonValue));
+      }
+      return fromNumber.apply(number);
+    }
+    throw new IllegalArgumentException("Invalid %s value: expected an integer or a string");
+  }
 
-    return value;
+  private static <C extends Collection<Object>> C toCqlCollection(
+      Column.ColumnType type, List<Object> jsonValues, Supplier<C> newInstance, C empty) {
+
+    if (jsonValues.isEmpty()) {
+      return empty;
+    }
+    Column.ColumnType elementType = type.parameters().get(0);
+    C result = newInstance.get();
+    int index = 0;
+    for (Object jsonValue : jsonValues) {
+      try {
+        result.add(toCqlValue(elementType, jsonValue));
+      } catch (Exception e) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Invalid %s element at index %d (%s)", type.rawType(), index, e.getMessage()));
+      }
+      index += 1;
+    }
+    return result;
   }
 
-  public static ValueModifier colToValue(Map.Entry<String, String> entry, Table tableData) {
-    String name = entry.getKey();
-    Column col = tableData.column(name);
-    Column.ColumnType type = col.type();
-    Object value = entry.getValue();
+  @SuppressWarnings("rawtypes")
+  private static Map<Object, Object> toCqlMap(Column.ColumnType type, List<Object> jsonEntries) {
+    if (jsonEntries.isEmpty()) {
+      return Collections.emptyMap();
+    }
+    Column.ColumnType keyType = type.parameters().get(0);
+    Column.ColumnType valueType = type.parameters().get(1);
+    LinkedHashMap<Object, Object> result = new LinkedHashMap<>();
+    int index = 0;
+    for (Object jsonValue : jsonEntries) {
+      if (!(jsonValue instanceof Map)
+          || !((Map) jsonValue).containsKey("key")
+          || !((Map) jsonValue).containsKey("value")) {
+        throw new IllegalArgumentException(
+            "Invalid Map value: inner elements must be objects with two fields 'key' and 'value'");
```

Comment:
Why must the fields be 'key' and 'value'?

Reply:
This is what a serialized map looks like:
```
[{"key":1,"value":"a"},{"key":2,"value":"b"}]
```
The entries are objects with two fields, these were the most logical names for them.

Alternatively, we could use two-element arrays, or allow arbitrary field names, but I don't like either of them because it's less clear and requires them to be in a given order.